File: FirstAkkaApplication/src/main/scala/akka/first/app/scala/actors/MasterActor.scala
package akka.first.app.scala.actors
import akka.actor.actorRef2Scala
import akka.actor.Actor
import akka.actor.ActorRef
import akka.actor.Props
import akka.first.app.scala.Result

class MasterActor extends Actor {
	
	val aggregateActor:ActorRef = context.actorOf(Props[AggregateActor],name="aggregate")
	val reduceActor:ActorRef = context.actorOf(Props(new ReduceActor(aggregateActor)),name="reduce")
	val mapActor:ActorRef = context.actorOf(Props(new MapActor(reduceActor)),name="map")

		def receive: Receive = {
		case message: String =>
			mapActor ! message
		case message:Result =>
			aggregateActor ! message
		case _ =>
	}
}

File: FirstAkkaApplication/src/main/java/akka/first/app/java/actors/MasterActor.java
package akka.first.app.java.actors;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.actor.UntypedActorFactory;
import akka.first.app.java.messages.Result;

public class MasterActor extends UntypedActor {

	private ActorRef aggregateActor = getContext().actorOf(
			new Props(AggregateActor.class), "aggregate");

	private ActorRef reduceActor = getContext().actorOf(
			new Props(new UntypedActorFactory() {
				public UntypedActor create() {
					return new ReduceActor(aggregateActor);
				}
			}), "reduce");

	private ActorRef mapActor = getContext().actorOf(
			new Props(new UntypedActorFactory() {
				public UntypedActor create() {
					return new MapActor(reduceActor);
				}
			}), "map");

	@Override
	public void onReceive(Object message) throws Exception {
		if (message instanceof String) {
			mapActor.tell(message);
		} else if (message instanceof Result) {
			aggregateActor.tell(message);
		} else
			unhandled(message);
	}
}
